timemachine/cpp/src/wrap_kernels.cpp

```diff
@@ -381,6 +381,173 @@ void declare_potential(py::module &m) {
             py::arg("params"),
             py::arg("box"),
             py::arg("lam"))
+        .def(
+            "execute_selective_batch",
+            [](timemachine::Potential &pot,
+               const py::array_t<double, py::array::c_style> &coords,
+               const py::array_t<double, py::array::c_style> &params,
+               const py::array_t<double, py::array::c_style> &boxes,
+               const py::array_t<double, py::array::c_style> &lambdas,
+               const bool compute_du_dx,
+               const bool compute_du_dp,
+               const bool compute_du_dl,
+               const bool compute_u) -> py::tuple {
+                if (coords.ndim() != 3 && boxes.ndim() != 3) {
+                    throw std::runtime_error("coords and boxes must have 3 dimensions");
+                }
+                if (coords.shape()[0] != boxes.shape()[0]) {
+                    throw std::runtime_error("number of batches of coords and boxes don't match");
+                }
+                if (params.ndim() < 2) {
+                    throw std::runtime_error("parameters must have at least 2 dimensions");
+                }
+                if (lambdas.size() < 1) {
+                    throw std::runtime_error("lambdas must have at least 1 value");
+                }
+                const long unsigned int coord_batches = coords.shape()[0];
+                const long unsigned int N = coords.shape()[1];
+                const long unsigned int D = coords.shape()[2];
+
+                const long unsigned int param_batches = params.shape()[0];
+                const long unsigned int P = params.size() / param_batches;
+
+                const long unsigned int lambda_batches = lambdas.size();
+
+                const long unsigned int K = coord_batches * param_batches * lambda_batches;
+
+                // initialize with fixed garbage values for debugging convenience (these should be overwritten by `execute_host`)
+                // Only initialize memory when needed, as buffers can be quite large
+                std::vector<unsigned long long> du_dx;
+                if (compute_du_dx) {
+                    du_dx.resize(K * N * D, 9999);
+                }
+                std::vector<unsigned long long> du_dp;
+                if (compute_du_dp) {
+                    du_dp.resize(K * P, 9999);
+                }
+                std::vector<unsigned long long> du_dl;
+                if (compute_du_dl) {
+                    du_dl.resize(K * N, 9999);
+                }
+                std::vector<unsigned long long> u;
+                if (compute_u) {
+                    u.resize(K * N, 9999);
+                }
```

Comment:
A little tripped up why these two are `K * N` rather than just `K`, but I think this is consistent the logic already present in the `execute` wrapper. May still be worth a line comment?

Reply:
Will add a comment